clients/js/src/utils.ts

```diff
@@ -98,27 +102,102 @@ function arrayifyParams(
   };
 }
 
-export async function prepareRecordRequest(
-  reqParams: AddRecordsParams | UpdateRecordsParams,
+export async function prepareRecordRequestForUpsert(
+  reqParams: UpsertRecordsParams | UpdateRecordsParams,
   embeddingFunction: IEmbeddingFunction,
-  update?: true,
 ): Promise<MultiRecordOperationParams> {
-  const { ids, embeddings, metadatas, documents } = arrayifyParams(reqParams);
+  const recordSet = arrayifyParams(reqParams);
+  const { ids, embeddings, documents } = recordSet;
+
+  if (!ids) {
+    throw new Error("ids cannot be undefined");
+  }
 
-  if (!embeddings && !documents && !update) {
+  if (!embeddings && !documents) {
     throw new Error("embeddings and documents cannot both be undefined");
   }
 
-  const embeddingsArray = embeddings
-    ? embeddings
-    : documents
-    ? await embeddingFunction.generate(documents)
-    : undefined;
+  validateIDs(ids);
+
+  recordSet.embeddings = await computeEmbeddings(
+    embeddingFunction,
+    embeddings,
+    documents,
+  );
+
+  return recordSet as MultiRecordOperationParams;
+}
+
+export async function prepareRecordRequestForUpdate(
+  reqParams: UpsertRecordsParams | UpdateRecordsParams,
+  embeddingFunction: IEmbeddingFunction,
+): Promise<MultiRecordOperationParams> {
+  const recordSet = arrayifyParams(reqParams);
+  const { ids, embeddings, documents } = recordSet;
+
+  if (!ids) {
+    throw new Error("ids cannot be undefined");
+  }
+
+  validateIDs(ids);
+
+  if (documents) {
+    recordSet.embeddings = await computeEmbeddings(
+      embeddingFunction,
+      embeddings,
+      documents,
+    );
+  }
+
+  return recordSet as MultiRecordOperationParams;
```

Comment:
```suggestion
  const recordSet: MultiRecordOperationParams = arrayifyParams(reqParams);
  const { ids, embeddings, documents } = recordSet;

  if (!ids) {
    throw new Error("ids cannot be undefined");
  }

  validateIDs(ids);

  if (documents) {
    recordSet.embeddings = await computeEmbeddings(
      embeddingFunction,
      embeddings,
      documents,
    );
  }

  return recordSet;
```

Reply:
`arrayifyParams` actually gives `MultiRecordOperationParamsWithIDsOptional` type instead of `MultiRecordOperationParams `